setup.sh

```diff
@@ -32,52 +36,54 @@ INSTALL_DIR=./yagit
 
 
 # ============================================================
+cd "$(dirname "$0")"
 mkdir -p build
 cd build
 
 # ============================================================
+DEPENDENCIES_PATHS=""
+TOOLCHAIN_FILE=""
 
-install () {
-    # $1 - path to library to install
+install_lib () {
+    # $1 - url to git repository of the library that will be installed
+    # $2 - tag or branch of the library
+    # $3 - installation mode (LOCAL or GLOBAL)
 
-    cd $1
-    mkdir -p build && cd build
-    cmake .. -DCMAKE_BUILD_TYPE=Release
-    cmake --build . --config Release -j
+    # extract repository name from url
+    repo_name_git=${1##*/}
+    repo_name=${repo_name_git%.git}
 
-    if [ $INSTALL_DEPENDENCIES == LOCAL ]; then
-        cmake --install . --prefix ./installed
-    elif [ $INSTALL_DEPENDENCIES == GLOBAL ]; then
-        sudo cmake --install .
+    if [ ! -d $repo_name ]; then
+        # clone git repo
+        git clone $1 -b $2 --depth 1 -c advice.detachedHead=false
     fi
 
-    cd ../..
-}
+    if [ ! -d $repo_name/build ]; then
+        cd $repo_name
+        mkdir -p build && cd build
 
-DEPENDENCIES_PATHS=""
-TOOLCHAIN_FILE=""
+        # configure and build
+        cmake .. -DCMAKE_BUILD_TYPE=Release
+        cmake --build . --config Release -j
 
-if [[ $INSTALL_DEPENDENCIES == LOCAL || $INSTALL_DEPENDENCIES == GLOBAL ]]; then
-    echo "INSTALLING DEPENDENCIES..."
-    mkdir -p deps && cd deps
+        # install
+        if [ $3 == LOCAL ]; then
+            cmake --install . --prefix ./installed
+        elif [ $3 == GLOBAL ]; then
+            sudo cmake --install .
+        fi
 
-    # GDCM
-    if [ ! -d GDCM ]; then
-        git clone https://github.com/malaterre/GDCM.git -b v3.0.22
-        install GDCM
+        cd ../..
     fi
+}
 
-    # xsimd
-    if [ ! -d xsimd ]; then
-        git clone https://github.com/xtensor-stack/xsimd.git -b 11.1.0
-        install xsimd
-    fi
+if [[ $INSTALL_DEPENDENCIES == LOCAL || $INSTALL_DEPENDENCIES == GLOBAL ]]; then
+    echo "INSTALLING DEPENDENCIES..."
+    mkdir -p deps && cd deps
 
-    # GoogleTest
-    if [ ! -d googletest ]; then
-        git clone https://github.com/google/googletest.git -b v1.13.0
-        install googletest
-    fi
+    install_lib https://github.com/malaterre/GDCM.git v3.0.22 $INSTALL_DEPENDENCIES
+    install_lib https://github.com/xtensor-stack/xsimd.git 11.1.0 $INSTALL_DEPENDENCIES
+    install_lib https://github.com/google/googletest.git v1.13.0 $INSTALL_DEPENDENCIES
```

Comment:
maybe you could bump this to a latest versions ?

Reply:
I plan to bump these to latest versions, but I think it would be better to do it in a separate commit with only this change with possible fixes in code due to these updates. Commit history will be more readable.